File: src/main/java/org/flossware/civilization/model/Technology.java
package org.flossware.civilization.model;

import java.util.List;
import java.util.Objects;

/**
 * Represents a technology node in the research tree.
 * Immutable record for thread safety and reproducibility.
 */
public record Technology(
    String id,
    String era,
    List<String> prerequisites,
    double researchCost,
    double diffusionRate
) {
    public Technology {
        Objects.requireNonNull(id, "Technology id cannot be null");
        Objects.requireNonNull(era, "Technology era cannot be null");
        Objects.requireNonNull(prerequisites, "Prerequisites cannot be null");

        if (researchCost <= 0) {
            throw new IllegalArgumentException("Research cost must be positive: " + researchCost);
        }
        if (diffusionRate < 0 || diffusionRate > 1) {
            throw new IllegalArgumentException("Diffusion rate must be in [0, 1]: " + diffusionRate);
        }

        prerequisites = List.copyOf(prerequisites);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/Technology.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable

@Serializable
data class Technology(
    val id: String,
    val era: String,
    val prerequisites: List<String>,
    val researchCost: Double,
    val diffusionRate: Double
) {
    init {
        require(researchCost > 0) { "Research cost must be positive: $researchCost" }
        require(diffusionRate in 0.0..1.0) { "Diffusion rate must be in [0, 1]: $diffusionRate" }
    }
}
